mod/quiz/edit.php

```diff
@@ -131,15 +130,33 @@
     redirect($afteractionurl);
 }
 
-if ((optional_param('addrandom', false, PARAM_BOOL)) && confirm_sesskey()) {
+if ((optional_param('addrandom', false, PARAM_BOOL) ||
+        optional_param('newcategory', false, PARAM_BOOL))
+        && confirm_sesskey()) {
+
     // Add random questions to the quiz.
     $structure->check_can_be_edited();
     $recurse = optional_param('recurse', 0, PARAM_BOOL);
     $addonpage = optional_param('addonpage', 0, PARAM_INT);
     $categoryid = required_param('categoryid', PARAM_INT);
     $randomcount = required_param('randomcount', PARAM_INT);
-    quiz_add_random_questions($quiz, $addonpage, $categoryid, $randomcount, $recurse);
 
+    // If we need to create new category.
+    if (optional_param('newcategory', false, PARAM_BOOL)) {
+        $qcobject = new \qbank_managecategories\question_category_object(
+            $pagevars['cpage'],
+            $thispageurl,
+            $contexts->having_one_edit_tab_cap('categories'),
+            $defaultcategoryobj->id,
+            $defaultcategory,
+            null,
+            $contexts->having_cap('moodle/question:add'));
+        $categoryname = required_param('categoryname', PARAM_TEXT);
+        $parentcategory = required_param('parentcategory', PARAM_TEXT);
+        $categoryid = $qcobject->add_category($parentcategory, $categoryname, '', true);
+    }
+
+    quiz_add_random_questions($quiz, $addonpage, $categoryid, $randomcount, $recurse);
```

Comment:
I forget removing this one, as I have replaced it with ws function mod_quiz_add_random_question